precompiles/utils/macro/tests/precompile/compile-fail/evm-data/arg-dont-impl-evmdata.stderr

```diff
@@ -0,0 +1,61 @@
+              (TupleElement0, TupleElement1, TupleElement2, TupleElement3, TupleElement4, TupleElement5, TupleElement6, TupleElement7)
+            and 27 others
+note: required by a bound in `EvmDataReader::<'a>::read`
+   --> $WORKSPACE/precompiles/utils/src/data/mod.rs
+    |
+    |     pub fn read<T: EvmData>(&mut self) -> MayRevert<T> {
+    |                    ^^^^^^^ required by this bound in `EvmDataReader::<'a>::read`
+
+error[E0277]: the trait bound `String: EvmData` is not satisfied
+   --> tests/precompile/compile-fail/evm-data/arg-dont-impl-evmdata.rs:23:43
+    |
+23  |     fn foo(test: &mut impl PrecompileHandle, arg: String) -> EvmResult {
+    |                                              ^^^ the trait `EvmData` is not implemented for `String`
+    |
+    = help: the following other types implement trait `EvmData`:
+              ()
+              (TupleElement0, TupleElement1)
+              (TupleElement0, TupleElement1, TupleElement2)
+              (TupleElement0, TupleElement1, TupleElement2, TupleElement3)
+              (TupleElement0, TupleElement1, TupleElement2, TupleElement3, TupleElement4)
+              (TupleElement0, TupleElement1, TupleElement2, TupleElement3, TupleElement4, TupleElement5)
+              (TupleElement0, TupleElement1, TupleElement2, TupleElement3, TupleElement4, TupleElement5, TupleElement6)
+              (TupleElement0, TupleElement1, TupleElement2, TupleElement3, TupleElement4, TupleElement5, TupleElement6, TupleElement7)
+            and 27 others
+note: required by a bound in `EvmDataWriter::write`
+   --> $WORKSPACE/precompiles/utils/src/data/mod.rs
+    |
+    |     pub fn write<T: EvmData>(mut self, value: T) -> Self {
+    |                     ^^^^^^^ required by this bound in `EvmDataWriter::write`
+
+error[E0277]: the trait bound `String: EvmData` is not satisfied
+  --> tests/precompile/compile-fail/evm-data/arg-dont-impl-evmdata.rs:23:5
+   |
+23 |     fn foo(test: &mut impl PrecompileHandle, arg: String) -> EvmResult {
+   |        ^^^ the trait `EvmData` is not implemented for `String`
+   |
+   = help: the following other types implement trait `EvmData`:
+             ()
+             (TupleElement0, TupleElement1)
+             (TupleElement0, TupleElement1, TupleElement2)
+             (TupleElement0, TupleElement1, TupleElement2, TupleElement3)
+             (TupleElement0, TupleElement1, TupleElement2, TupleElement3, TupleElement4)
+             (TupleElement0, TupleElement1, TupleElement2, TupleElement3, TupleElement4, TupleElement5)
+             (TupleElement0, TupleElement1, TupleElement2, TupleElement3, TupleElement4, TupleElement5, TupleElement6)
+             (TupleElement0, TupleElement1, TupleElement2, TupleElement3, TupleElement4, TupleElement5, TupleElement6, TupleElement7)
+           and 27 others
+   = note: required because of the requirements on the impl of `EvmData` for `(String,)`
```

Comment:
Is this a test or is this the compiler output upon this error? If it is the compiler output, the repetitive parts should be taken out
```suggestion
```

Reply:
This is a compiler output error.

However it occurs **after** the successful expansion because `String` don't implement `EvmData`, which cannot be known in a macro. Thus the goal here is to give to the compile a Span that makes errors usable for the developers. The fact that the type implements `EvmData` is used 3 times, hense the 3 slightly different errors (read, write, tuple).

Here it properly refers to the argument that doesn't implement `EvmData`, and only the function name in the last error (since here it is the tuple `(String,)` as whole that indirectly don't implement `EvmData`, so I can't find a way to give a more precise Span.

Reply:
Ok sounds good